presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java

```
@@ -99,38 +100,130 @@ public static BindableAggregationFunction generateAggregationBindableFunction(Cl
                         builder.add(
                                 new BindableAggregationFunction(
-                                        signature,
-                                        new AggregationHeader(getDescription(aggregationDefinition, outputFunction),
-                                                aggregationAnnotation.approximate(),
-                                                aggregationAnnotation.decomposable()),
-                                        new AggregationImplementations(ImmutableMap.of(signature, onlyImplementation),
+                                        onlyImplementation.getSignature(),
+                                        header,
+                                        new AggregationImplementations(ImmutableMap.of(),
                                                 ImmutableList.of(),
-                                                ImmutableList.of())));
+                                                ImmutableList.of(onlyImplementation))));
+                    }
+                }
+            }
+        }
+
+        return builder.build();
+    }
+
+    public static List<BindableAggregationFunction> generateBindableAggregationFunction(Class<?> aggregationDefinition)
+    {
+        ImmutableMap.Builder<Signature, AggregationImplementation> exactImplementations = ImmutableMap.builder();
+        ImmutableList.Builder<AggregationImplementation> specializedImplementations = ImmutableList.builder();
+        ImmutableList.Builder<AggregationImplementation> genericImplementations = ImmutableList.builder();
+        Optional<Signature> genericSignature = Optional.empty();
+        AggregationHeader header = parseHeader(aggregationDefinition);
+
+        for (Class<?> stateClass : getStateClasses(aggregationDefinition)) {
+            for (Method outputFunction : getOutputFunctions(aggregationDefinition, stateClass)) {
```

getOutputFunctions(def, state) and getInputFunctions(def, state) should be called getOutputFunction and getInputFunction and return Optional<...>.

---

In newer version there is a check by getOnlyElement method to ensure that there is exactly one function like that. We want to fail fast if functions are erratically annotated so no optional here - we just return all we have and throw an exception if it doesn't meet our expectations.